evm_loader/faucet/src/solana.rs

```diff
@@ -44,117 +43,165 @@ pub fn convert_whole_to_fractions(amount: u64) -> Result<u64> {
 /// When in_fractions == true, amount is treated as amount in galans (10E-9).
 pub async fn transfer_token(
     id: &str,
-    signer: Keypair,
+    signer: Vec<u8>,
     ether_address: ethereum::Address,
     amount: u64,
     in_fractions: bool,
 ) -> Result<()> {
     let evm_loader_id = Pubkey::from_str(&config::solana_evm_loader()).wrap_err_with(|| {
         format!(
-            "{} config::solana_evm_loader returns {}",
-            id,
+            "config::solana_evm_loader returns {}",
             &config::solana_evm_loader()
         )
     })?;
     let token_mint_id = Pubkey::from_str(&config::solana_token_mint_id()).wrap_err_with(|| {
         format!(
-            "{} config::solana_token_mint_id returns {}",
-            id,
+            "config::solana_token_mint_id returns {}",
             &config::solana_token_mint_id(),
         )
     })?;
 
-    let signer_account = signer.pubkey();
+    let signer_account = Keypair::from_bytes(&signer)?.pubkey();
     let signer_token_account =
         spl_associated_token_account::get_associated_token_address(&signer_account, &token_mint_id);
 
     let (account, _nonce) = make_solana_program_address(&ether_address, &evm_loader_id);
     let token_account =
         spl_associated_token_account::get_associated_token_address(&account, &token_mint_id);
 
-    let id = id.to_owned();
-    let r = tokio::task::spawn_blocking(move || -> Result<()> {
-        let client = get_client();
-        let mut instructions = vec![];
+    info!("{} spl_token id = {}", id, spl_token::id());
+    info!("{} signer_token_account = {}", id, signer_token_account);
+    info!("{} token_mint_id = {}", id, token_mint_id);
+    info!("{} token_account = {}", id, token_account);
+    info!("{} signer_account = {}", id, signer_account);
+    info!("{} amount = {}", id, amount);
+    info!(
+        "{} token_decimals = {}",
+        id,
+        config::solana_token_mint_decimals()
+    );
+
+    let exists = ether_account_exists(id.to_owned(), account, token_account).await?;
+    if !exists {
+        create_ether_account(signer.clone(), signer_account, evm_loader_id, ether_address).await?;
```

Comment:
Why do you need to separate one transaction to different for creating an account and transfer tokens?
As I understand in this case faucet will work for 26 seconds for the request. Moreover, it transfers tokens only after 13 seconds after request in a successful case!

Reply:
This is the point. The instruction EvmInstruction::CreateAccount fails in some cases: "account already in use".